src/fiber/graph.rs

```diff
@@ -0,0 +1,1174 @@
+use super::network::{get_chain_hash, SendPaymentCommand};
+use super::path::NodeHeap;
+use super::types::Pubkey;
+use super::types::{ChannelAnnouncement, ChannelUpdate, Hash256, NodeAnnouncement};
+use crate::fiber::path::{NodeHeapElement, ProbabilityEvaluator};
+use crate::fiber::types::OnionInfo;
+use crate::invoice::CkbInvoice;
+use ckb_types::packed::OutPoint;
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+use std::collections::HashMap;
+use tentacle::multiaddr::Multiaddr;
+use tentacle::secio::PeerId;
+use thiserror::Error;
+use tracing::log::error;
+use tracing::{debug, info, warn};
+
+const DEFAULT_MIN_PROBABILITY: f64 = 0.01;
+
+#[serde_as]
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+/// Details about a node in the network, known from the network announcement.
+pub struct NodeInfo {
+    pub node_id: Pubkey,
+
+    // The time when the node was last updated. This is the time of processing the message,
+    // not the time of the NodeAnnouncement itself.
+    pub timestamp: u64,
+
+    pub anouncement_msg: NodeAnnouncement,
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct ChannelInfo {
+    pub funding_tx_block_number: u64,
+    pub funding_tx_index: u32,
+    pub announcement_msg: ChannelAnnouncement,
+    // The last update (with the timestamp) for the channel received from the network.
+    pub one_to_two: Option<(ChannelUpdateInfo, u64)>,
+    pub two_to_one: Option<(ChannelUpdateInfo, u64)>,
```

Comment:
It's better in readability to add a timestamp field in `ChannelUpdateInfo`?

Reply:
@contrun

Reply:
Yes. I added that.